pkg/serviceaccount/claims.go

```diff
@@ -76,3 +79,102 @@ func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirati
 	}
 	return sc, pc
 }
+
+func NewValidator(audiences []string, lookup bool, getter ServiceAccountTokenGetter) Validator {
+	return &validator{
+		auds:   audiences,
+		lookup: lookup,
+		getter: getter,
+	}
+}
+
+type validator struct {
+	auds   []string
+	lookup bool
+	getter ServiceAccountTokenGetter
+}
+
+var _ = Validator(&validator{})
+
+func (v *validator) Validate(_ string, public *jwt.Claims, privateObj interface{}) (string, string, string, error) {
+	private, ok := privateObj.(*privateClaims)
+	if !ok {
+		glog.Errorf("jwt validator expected private claim of type *privateClaims but got: %T", privateObj)
+		return "", "", "", errors.New("Token could not be validated.")
+	}
+	err := public.Validate(jwt.Expected{
+		Audience: v.auds,
+		Time:     now(),
+	})
+	switch {
+	case err == nil:
+	case err == jwt.ErrExpired:
+		return "", "", "", errors.New("Token has expired.")
+	case err == jwt.ErrInvalidAudience:
+		return "", "", "", errors.New("Token is invalid for this audience.")
+	default:
+		glog.Errorf("unexpected validation error: %T", err)
+		return "", "", "", errors.New("Token could not be validated.")
+	}
+
+	if v.lookup {
+		namespace := private.Kubernetes.Namespace
+		saref := private.Kubernetes.Svcacct
+		podref := private.Kubernetes.Pod
+		secref := private.Kubernetes.Secret
+		// Make sure service account still exists (name and UID)
+		serviceAccount, err := v.getter.GetServiceAccount(namespace, saref.Name)
+		if err != nil {
+			glog.V(4).Infof("Could not retrieve service account %s/%s: %v", namespace, saref.Name, err)
+			return "", "", "", err
+		}
+		if serviceAccount.DeletionTimestamp != nil {
+			glog.V(4).Infof("Service account has been deleted %s/%s", namespace, saref.Name)
+			return "", "", "", fmt.Errorf("ServiceAccount %s/%s has been deleted", namespace, saref.Name)
+		}
+		if string(serviceAccount.UID) != saref.UID {
+			glog.V(4).Infof("Service account UID no longer matches %s/%s: %q != %q", namespace, saref.Name, string(serviceAccount.UID), saref.UID)
+			return "", "", "", fmt.Errorf("ServiceAccount UID (%s) does not match claim (%s)", serviceAccount.UID, saref.UID)
+		}
+
+		if secref != nil {
+			// Make sure token hasn't been invalidated by deletion of the secret
+			secret, err := v.getter.GetSecret(namespace, secref.Name)
+			if err != nil {
+				glog.V(4).Infof("Could not retrieve token %s/%s for service account %s/%s: %v", namespace, secref.Name, namespace, saref.Name, err)
```

Comment:
nit, update messages to reflect the object that couldn't be retrieved (s/token/secret/, and below s/token/serviceaccount/)